### ISSUE TITLE
Can't cloudpickle attrs classes

### ISSUE THREAD

**Issue by danpf** [python-attrs/attrs#458]
I brought this up on stackoverflow and @hynek  said that the quick fix is to set 'repr=False', but I very much like the repr features of attrs classes, so I would be sad to lose that.

A detailed description of the problem is at:
https://stackoverflow.com/questions/53073440/how-to-pickle-python-attrs-class-functions-via-cloudpickle

but briefly:

This works:
```
import pickle
import attr

@attr.s(auto_attribs=True)
class myclass:
    an_int: int
    a_str: str
    a_float: float

    def myfunc(self):
        return f"{self.an_int} + {self.a_str} + {self.a_float}"

mc = myclass(1, "test", 2.4)
f = pickle.loads(pickle.dumps(mc.myfunc))
print(f())
```

This doesn't and gives me the error: 
`gives me the error: TypeError: can't pickle _thread._local objects`
```
import attr
import cloudpickle

@attr.s(auto_attribs=True)
class myclass:
    an_int: int
    a_str: str
    a_float: float

    def myfunc(self):
        return f"{self.an_int} + {self.a_str} + {self.a_float}"

mc = myclass(1, "test", 2.4)
f = cloudpickle.loads(cloudpickle.dumps(mc.myfunc))
print(f())
```

**Comment by wsanchez** [comment]
Adding @hynek's comment from SO, since that's the actual issue here:

This is caused by attrs using thread locals for breaking up object cycles in __repr__ and cloudpickle not liking that. For reference, it was introduced in https://github.com/python-attrs/attrs/pull/358.

**Comment by egberts** [comment]
That looks like my #493, no?

**Comment by Lingepumpe** [comment]
This seems worse now, even with repr=False I cannot run cloudpickle.dumps on @attrs classes:
```
from attr import attrs, attrib
import cloudpickle

@attrs (repr=False)
class Foo:
    x: int = attrib()

cloudpickle.dumps(Foo)
```
Fails with `TypeError: cannot pickle 'property' object` for me.
python 3.8.1, attrs = "19.3.0", cloudpickle = "1.2.2"

**Comment by hynek** [comment]
The bug mentioned by @Lingepumpe seems to be fixed, what's left is the issue with `repr()`.

**Comment by JarnoRFB** [comment]
This is indeed unfortunate if one wants to use an `attrs` heavy code base with [ray](https://github.com/ray-project/ray) which relies on `cloudpickle`. I could image that it is also a problem with other projects that utilize `cloudpickle` for doing some kind of cluster / cloud computing. Would fixing this require changing the way of breaking recursion error in `repr` or would it also be an option to make the use of `_thread._local` optional, so maybe set it only if one expects recursive references?

**Comment by hynek** [comment]
I would love for attrs to work with cloudpickle but having to set an obscure option that is added only for that purpose doesn't seem a good way to achieve that.

In the end, you can always set `repr=False` and be done with it.

---

A better solution I think would be if cloudpickle could be respectfully convinced to add support for attrs and I'd be happy to help out to get there.

**Comment by pwais** [comment]
I hypothesize there might be a fix for this issue in `attrs` by making the global variable `_already_repring` not thread-local.  (original discussion https://github.com/python-attrs/attrs/issues/754#issuecomment-773100398 ).  It's not clear to me that the variable needed to be thread-local in the first place ( https://github.com/python-attrs/attrs/pull/358 ).  I'm not sure how best to test / experiment though; I imagine a change here could impact a lot of things downstream.

Workaround: If you put your `attrs`-based class in a separate file / python module, then I've found both `pickle` and `cloudpickle` seem to be OK with that.  I'm not sure how to include python code with Apache Ray, but for Apache Spark there's a well-defined API.  For jupyter notebooks + Spark you can use `%%writefile` to create a python module on-disk that gets included in your Spark job.  I have a demo here: https://pwais.github.io/oarph-oarphs/2020/02/03/atrick.html

**Comment by hynek** [comment]
My understanding is that the reason is:

1. we need global state to keep track of what is tracked
2. if two threads tried to repr the same instance, you get a race condition.

Traditionally you'd use a lock or mutex for something like that, but I think a thread-local variable is more elegant and locks/mutexes are also unlikely to be pickleable?

Race conditions are hellish hard to test, so I don't think we can come up with a unit test here.

***

I'm afraid I don't quite understand what you mean by the separate file exactly. Especially because the thread-local variable that's choking cloud pickle is within attrs itself?

**Comment by RonnyPfannschmidt** [comment]
cloudpickle seems to try and transfer all code and related details in a pickle it should add attrs support itself

**Comment by Tinche** [comment]
We could maybe stick the threadlocal onto the class itself? Would that help?

**Comment by hynek** [comment]
I mean there's issue #320 since January 2020, but nobody seems to have the time/motivation to fix it. I find it unfortunate but I can't do more than offer my help to them fixing it. I can't run around and fix everybody's code or adapt my code to everybody's needs.

**Comment by hynek** [comment]
fixed by #857!
